packages/next/next-server/server/next-server.ts

```diff
@@ -315,6 +323,76 @@ export default class Server {
       },
     ]
 
+    let rewrites: {
+      source: string
+      destination: string
+    }[] = []
+    let redirects: {
+      source: string
+      destination: string
+      statusCode?: number
+    }[] = []
+
+    if (this.dev) {
+      if (typeof this.nextConfig.rewrites === 'function') {
+        rewrites = await this.nextConfig.rewrites()
+      }
+      if (typeof this.nextConfig.redirects === 'function') {
+        redirects = await this.nextConfig.redirects()
+      }
+    } else {
+      // use built routes
+      const routesManifest = require(join(this.distDir, ROUTES_MANIFEST))
+      rewrites = routesManifest.rewrites
+      redirects = routesManifest.redirects
+    }
+
+    routes.push(
+      ...rewrites.map(rewrite => {
+        return {
+          match: route(rewrite.source),
+          fn: async (req, res, params, parsedUrl) => {
+            // TODO: investigate limiting rewrite depth
+            // to prevent accidental infinite rewrites
+            let destination = rewrite.destination
+
+            for (const key of Object.keys(params)) {
+              const param = params[key]
+              destination = destination.replace(
+                new RegExp(`:${key}`, 'g'),
+                param
+              )
+            }
+            const newParsedUrl = parseUrl(destination, true)
+            let newParams: { [name: string]: string } | false = false
+
+            const nextMatch = this.router.routes.find(route => {
+              newParams = route.match(rewrite.destination)
+              if (newParams) {
+                return true
+              }
+            })
+            // we should always find a match because we have a final catch all
+            // which will render 404 for us
+            return nextMatch!.fn(req, res, newParams as any, newParsedUrl)
+          },
+        } as Route
+      })
+    )
+
+    routes.push(
+      ...redirects.map(redirect => {
```

Comment:
Redirects should be a single round trip in-case of chaining, we need to probably handle this differently. 🤔

Reply:
Updated, chaining is handled for both in a similar way now